[13804] Topic, instances and keys (#324)
eProsima/Fast-DDS-docs: docs/fastdds/dds_layer/topic/instances.rst

.. include:: ../../../03-exports/aliases-api.include

.. _dds_layer_topic_instances:

Topics, keys and instances
==========================

> Even though, by definition, a Topic corresponds to a single data type, it is possible that several topic instances may

Reviewer: Suggestion:

A Topic is linked to a data type and each data sample related to a Topic could be seen as an update in the information described by the data type. But it is possible to add a logical separation and have several instances of a Topic, and a data sample should be an update of the information about a concrete  instance.

The diagram below shows the statechart of |SampleInfo::instance_state-api| and |SampleInfo::view_state-api| for a single
instance.

.. figure:: /01-figures/instance-lifecycle.png
    :align: center

.. _dds_layer_topic_instance_examples:

Practical applications
----------------------

This section provides a couple of examples to help clarify the use of DDS instances.

For example, instances could be used to track commercial flights.
The key identifying the instances could be the airline name and the flight number, while the sample data would be the
location of each flight being tracked at any given time.
Instead of creating a DataReader and Topic each time a flight takes off, the same control tower DataReader could manage
the information published by each plane DataWriter in the same Topic using this DDS standard feature.
Once the flight has landed, the instance can be disposed (no updates are expected); and if the flight is returned to the
hangar, the instance is unregistered.

The following IDL may define the data model of this specific example:

.. code-block:: idl

    struct FlightPosition
    {
        // Unique ID: airline name

>         @key
>         string<256> airline_name;

Reviewer 2: ```suggestion
        @key string<256> airline_name;
```

        // Unique ID: flight number

>         int flight_number;

Reviewer 2: I'd put this as part of the key as well

        // Coordinates
        double latitude;
        double longitude;
        double altitude;
    };

Then, writing to an specific instance will be done as below:

> .. literalinclude:: /../code/DDSCodeTester.cpp

Reviewer 2: Please de-indent

.. include:: ../../../03-exports/aliases-api.include

.. _dds_layer_topic_instances:

Topics, keys and instances
==========================

> Even though, by definition, a Topic corresponds to a single data type, it is possible that several topics may refer to

Reviewer 2: ```suggestion
Even though, by definition, a Topic corresponds to a single data type, it is possible that several topic instances may refer to
```

the same data type.
Therefore, a Topic identifies data of a single type, ranging from one single instance to a whole collection of instances
of that given type, as shown in the figure below.

.. figure:: /01-figures/instances.png
    :align: center

> The different instances gathered under the same topic are distinguishable by means of some data fields that form the key

Reviewer 2: ```suggestion
The different instances gathered under the same topic are distinguishable by means of one or more data fields that form the key
```

to that data set.
The key description has to be indicated to the middleware.

> The rule is simple: different data values with the same key value represent successive values for the same instance,

Reviewer 2: ```suggestion
The rule is simple: different data values with the same key value represent successive data samples for the same instance,
```

> while different data values with different keys represent different instances.

Reviewer 2: ```suggestion
while different data values with different keys represent different topic instances.
```

If no key is provided, the data set associated with the Topic is restricted to a single instance.
Please refer to :ref:`dds_layer_topic_keyed_data_types` for more information about how to set the key in
*eProsima Fast DDS*.

> 

Reviewer 2: I'dd add a title `Instance advantages`

The advantage of using instances instead of creating a new :ref:`dds_layer_publisher_dataWriter`,
:ref:`dds_layer_subscriber_dataReader`, and :ref:`dds_layer_topic_topic` is that the corresponding entity is already
created and discovered.

> Consequently, there is less memory usage and no new discovery (with the related metatraffic involved as explained in

Reviewer 2: ```suggestion
Consequently, there is less memory usage, and no new discovery (with the related metatraffic involved as explained in
```

:ref:`discovery`) is necessary.

> Another advantage is that some QoS are applied per-instance.
> For example, the :ref:`historyqospolicy` is kept for each instance in the DataWriter.

Reviewer 2: ```suggestion
Another advantage is that several QoS are applied per Topic instance; e.g. the :ref:`historyqospolicy` is kept for each instance in the DataWriter.
```

Thus, instances could be tuned to a wide range of applications.

> 

Reviewer 2: Move examples to separate sections after lifecycle

For example, instances could be used to track commercial flights.

> The key identifying the instances could be the airline name and the flight number and the sample the location of each

Reviewer 2: ```suggestion
The key identifying the instances could be the airline name and the flight number, while the sample data would be the location of each
```

> flight being tracked.

Reviewer 2: ```suggestion
flight being tracked at any given time.
```

> Instead of creating a DataWriter, DataReader and Topic each time a flight takes off, the same DDS entities may be reused

Reviewer 2: I'd elaborate that the plane would have the DataWriter and the control tower a single DataReader managing several instances (one per flight)

using this DDS standard feature.
Once the flight has landed, the instance can be disposed (no updates are expected); and if the flight is returned to the
hangar, the instance is unregistered.

Instances could also be used as a relational database.

> The instance key is analogous to the primary key in a database (unique identifier of something within your data).

Reviewer 2: ```suggestion
The instance key is analogous to the primary key in the database (unique identifier of something within the data).
```

> Writing a new sample to the instance is similar to insert/update the database; and disposing the instance is like

Reviewer 2: ```suggestion
In this scenario, writing a new sample to the instance is similar to insert/update the database; and disposing the instance is like
```

> removing the data corresponding to the given primary key from the database.

Reviewer 2: ```suggestion
deleting the data corresponding to the given primary key from the database.
```

.. literalinclude:: /../code/DDSCodeTester.cpp
   :language: c++

>    :dedent: 8

Reviewer 2: Please look at the generated HTML
```suggestion
   :dedent: 12
```
